corehq/apps/reports/views.py

```diff
@@ -1039,40 +1036,56 @@ def post(self, request, *args, **kwargs):
         return self.get(request, *args, **kwargs)
 
 
-class ReportNotificationUnsubscribeView(LoginAndDomainMixin, TemplateView):
+class ReportNotificationUnsubscribeView(TemplateView):
     template_name = 'reports/notification_unsubscribe.html'
     urlname = 'notification_unsubscribe'
     report = None
 
     @use_bootstrap3
     def get(self, request, *args, **kwargs):
-        self.domain = request.domain
         try:
             self.report = ReportNotification.get(kwargs.pop('scheduled_report_id'))
-            user_email = request.couch_user.get_email()
-            if user_email not in self.report.all_recipient_emails:
-                messages.error(request, _('You are already unsubscribed from this report.'))
-            else:
-                return super(ReportNotificationUnsubscribeView, self).get(request, *args, **kwargs)
+            email = kwargs.pop('user_email')
+
+            if kwargs.pop('scheduled_report_secret') != self.report.get_secret(email):
+                raise ValidationError('Invalid link')
+            if email not in self.report.all_recipient_emails:
+                raise ValidationError('This email address has already been unsubscribed.')
+            return super(ReportNotificationUnsubscribeView, self).get(request, *args, **kwargs)
+        except ValidationError as err:
+            messages.error(request, _(err.message))
         except ResourceNotFound:
-            pass
+            messages.error(request, _('Could not find the requested Scheduled Report'))
```

Comment:
If the user isn't logged in they don't see these messages because the home page doesn't support messages. I think you should return to the same page and display the message there.